### PR TITLE
Improve lookup table creation and type inference #1235

### PR DESCRIPTION
**Pull request by boryanagoncharenko**
**Description**

- separate lookup table creation in 2:
  in the first pass we only keep the name of the entry and the Tree object of its value
  in the second pass we infer the type of the lookup entry

- changed list access index to be 1-based since level 2

- refactor hedy.transpile_inner method

- fixed a translation error in the adventures/el.yaml causing a flaky test

**Fix for**

#1235 

**How to test**

- All changes should be fully backwards compatible, so ALL tests must pass.

### CODE REVIEW

File: hedy.py [hunk 1]
```diff
@@ -568,16 +645,13 @@ def ask(self, args):
         argument = process_characters_needing_escape(args[0])
         return "answer = input('" + argument + "')"
 
-    def forward(self,args):
+    def forward(self, args):
```
Comment (Felienne) [inline hedy.py:648]: Love these tiny styling fixes!! ❤️

**Review by Felienne — CHANGES_REQUESTED** [review]
Only tiny changes in the test, the rest looks amazing to me!

File: hedy.py [hunk 2]
```diff
@@ -189,188 +212,231 @@ def list_access(self, args):
...
+        self.add_to_lookup(iterator, trimmed_tree)
+
+    def for_loop(self, tree):
+        iterator = str(tree.children[0])
+        # the tree is trimmed to skip contain the inner commands of the loop since
+        # they are not needed to infer the type of the iterator variable
+        trimmed_tree = Tree(tree.data, tree.children[0:3], tree.meta)
+        self.add_to_lookup(iterator, trimmed_tree)
+
+    def add_to_lookup(self, name, tree, skip_hashing=False):
```
Comment (Felienne) [inline hedy.py:278]: love the extra argument to indicate hashing, a lot less error-prone and more clear this way

**Review by Felienne — CHANGES_REQUESTED** [review]
One more remark, need to make a separate review it seems

File: grammars/level2-Additions.lark [hunk 3]
```diff
@@ -16,7 +16,7 @@ text: /([^\r\n,!?.]+)/ -> text
 text_ask: /([^\r\n!?.]+)/ -> text //as ask may contain commas (punctionation is still needed separately so they can be printed after vars)
 text_list: /([^\r\n,]+)/ -> text // list elements may contain punctionationm but not commas or course, these are separators
 
-index : DIGIT+
+index : INT //todo FH-June 21 : why does this need its own rule? can't it just be a token
```
[the remark this answers is not in the document]
Comment (boryanagoncharenko) [reply grammars/level2-Additions.lark:19]: I will try removing the index rule and see what happens when it is just a token

[the remark this answers is not in the document]
Comment (Felienne) [reply grammars/level2-Additions.lark:19]: thanks for looking into this! feel a bit bad about leaving crap like that in the code base now, sorry!

File: tests/test_level_11.py [hunk 4]
```diff
@@ -160,6 +160,64 @@ def test_assign_string_with_quotes_and_string_value(self):
     self.assertEqual(expected, result.code)
     self.assertEqual(False, result.has_turtle)
 
+  def test_print_multiple_calcs(self):
```
[the remark this answers is not in the document]
Comment (boryanagoncharenko) [reply tests/test_level_11.py]: thanks for pointing it out! I revised all tests in level 11.

File: grammars/level2-Additions.lark (hunk 3 above)
[the remark this answers is not in the document]
Comment (boryanagoncharenko) [reply grammars/level2-Additions.lark]: It seems that no tests are broken when there is no separate index rule. So it is gone now :)

**Review by Felienne — APPROVED** [review]

**Pull request merged**